File: binary-search/search_2D_matrix.py
"""
You are given an m x n 2-D integer array matrix and an integer target.

    Each row in matrix is sorted in non-decreasing order.
    The first integer of every row is greater than the last integer of the previous row.

Return true if target exists within matrix or false otherwise.

Can you write a solution that runs in O(log(m * n)) time?

Example 1:
Input: matrix = [[1,2,4,8],[10,11,12,13],[14,20,30,40]], target = 10
Output: true

Example 2:
Input: matrix = [[1,2,4,8],[10,11,12,13],[14,20,30,40]], target = 15
Output: false

Constraints:
    m == matrix.length
    n == matrix[i].length
    1 <= m, n <= 100
    -10000 <= matrix[i][j], target <= 10000
"""


class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        left, right = 0, len(matrix) - 1
        while left <= right:
            mid = (left + right) // 2
            if target > matrix[mid][-1]:
                left = mid + 1
            elif target < matrix[mid][0]:
                right = mid - 1
            else:
                break

        if not (left <= right):
            return False

        i = mid
        left, right = 0, len(matrix[i]) - 1
        while left <= right:
            mid = (left + right) // 2
            if target == matrix[i][mid]:
                return True
            elif target < matrix[i][mid]:
                right = mid - 1
            else:
                left = mid + 1

        return False

File: binary-search/search_2D_matrix.zig
const std = @import("std");
const assert = std.debug.assert;
const testing = std.testing;
const Allocator = std.mem.Allocator;

const Solution = struct {
    fn search_matrix(matrix: []const []const i64, target: i64) bool {
        _ = matrix;
        _ = target;
        return true;
    }
};

test "test solution" {
    var matrix = [_][4]i64{
        .{ 1, 2, 4, 8 },
        .{ 10, 11, 12, 13 },
        .{ 14, 20, 30, 40 },
    };

    var rows = [_][]const i64{
        &matrix[0],
        &matrix[1],
        &matrix[2],
    };
    
    const target = 10;
    const solution = Solution.search_matrix(&rows, target);
    try testing.expectEqual(true, solution);
}
